scripts/chow_ni_tomography/step3_combine.py

```diff
@@ -0,0 +1,153 @@
+import numpy as np
+import pandas as pd
+from qcore import geo
+from shared import (
+    write_file,
+    lats,
+    lons,
+    chow_elevs,
+    v_file_template,
+    step1_outdir,
+    step2_outdir,
+    step3_outdir,
+    log_dir,
+    MAX_DIST,
+    v_types,
+)
+
+
+step3_outdir.mkdir(parents=True, exist_ok=True)
+log_dir.mkdir(parents=True, exist_ok=True)
+
+new_chow_elevs = [-750, -620] + chow_elevs + [15]  # add missing elevs to chow_elevs
+
+
+def nonzero_grid(v_array):
+    (
+        nonzero_vert_ids,
+        nonzero_hori_ids,
+    ) = v_array.nonzero()  # return row col indices of nonzero values
+    return nonzero_vert_ids, nonzero_hori_ids
+
+
+#
+def find_boundaries(nonzero_vert_ids, nonzero_hori_ids):
+
+    right_boundary = dict(
+        zip(nonzero_vert_ids, nonzero_hori_ids)
+    )  # for a given row index, the maximum col index is stored
+    left_boundary = dict(
+        zip(nonzero_vert_ids[::-1], nonzero_hori_ids[::-1])
+    )  # for a given row index, the minimum col index is stored
+
+    top_boundary = dict(
+        zip(nonzero_hori_ids, nonzero_vert_ids)
+    )  # for a given col index, x, the maximum row index is stored
+    bottom_boundary = dict(
+        zip(nonzero_hori_ids[::-1], nonzero_vert_ids[::-1])
+    )  # for a given col index, the minimum index is stored
+
+    return top_boundary, bottom_boundary, right_boundary, left_boundary
+
+
+# Find the distance to the 4 boundaries from a location, and return the minimum of 4 distances.
+def distance_from_boundary(x_id, y_id, boundaries):
+    top_boundary, bottom_boundary, right_boundary, left_boundary = boundaries
+    res = geo.get_distances(
+        np.array(
+            [
+                [lats[x_id], lons[left_boundary[x_id]]],
+                [lats[x_id], lons[right_boundary[x_id]]],
+                [lats[top_boundary[y_id]], lons[y_id]],
+                [lats[bottom_boundary[y_id]], lons[y_id]],
+            ]
+        ),
+        lats[x_id],
+        lons[y_id],
+    )
+    return min(res)
+
+
+def combine_and_smoothe(
+    nz2020_tomo_elevs_ext_array,
+    chow_tomo_grid_mapped_array,
+    nonzero_vert_ids,
+    nonzero_hori_ids,
+    boundaries,
+    logfile=None,
+):
+
+    final_v_array = (
+        nz2020_tomo_elevs_ext_array.copy()
+    )  # start with nz2020_tomo_elevs_ext_array
+
+    with open(logfile, "w") as f:
+        for i in range(
+            len(nonzero_vert_ids)
+        ):  # go through non_zero values (ie. points in chow NI domain)
+            lat_id = nonzero_vert_ids[i]
+            lon_id = nonzero_hori_ids[i]
+
+            d = distance_from_boundary(lat_id, lon_id, boundaries)
+
+            # if beyond MAX_DIST, use values from Chow tomo, otherwise, smoothe.
```

Comment:
Done. English seems to have evolved since I learned it.
<img width="870" alt="Screen Shot 2022-03-10 at 11 11 01 PM" src="https://user-images.githubusercontent.com/466989/157679304-303c0ee7-23f7-4fe9-ab5e-03d088685893.png">